dsn_test.go

```diff
@@ -879,3 +886,142 @@ func TestDSN(t *testing.T) {
+	pemData := pem.EncodeToMemory(
+		&pem.Block{
+			Type:  "RSA PRIVATE KEY",
+			Bytes: pemKey,
+		},
+	)
+	keyFile := createTmpFile("exampleKey.pem", pemData)
+	defer os.Remove(keyFile)
+
+	parsedKey, err := parsePrivateKeyFromFile(keyFile)
+	if err != nil {
+		t.Errorf("unable to parse pam file from path: %v, err: %v", keyFile, err)
+	} else if !parsedKey.Equal(generatedKey) {
+		t.Errorf("generated key does not equal to parsed key from file\ngeneratedKey=%v\nparsedKey=%v",
+			generatedKey, parsedKey)
+	}
+}
+
+func createTmpFile(fileName string, content []byte) string {
+	tempFile, _ := os.CreateTemp("", fileName)
+	tempFile.Write(content)
+	absolutePath := tempFile.Name()
+	return absolutePath
+}
+
+func TestGetFromEnv(t *testing.T) {
+	os.Setenv("SF_TEST", "test")
+	defer os.Unsetenv("SF_TEST")
+	result, err := GetFromEnv("SF_TEST", true)
+
+	if err != nil {
+		t.Error("failed to read SF_TEST environment variable")
+	}
+	if result != "test" {
+		t.Errorf("Incorrect value read for SF_TEST. Expected: test, read %v", result)
+	}
+}
+
+func TestGetFromEnvFailOnMissing(t *testing.T) {
+	_, err := GetFromEnv("SF_TEST_MISSING", true)
+	if err == nil {
+		t.Error("should report error when there is missing env parameter")
+	}
+}
+
+type configParamToValue struct {
+	configParam string
+	value       string
+}
+
+func TestGetConfigFromEnv(t *testing.T) {
+	envMap := map[string]configParamToValue{
+		"SF_TEST_ACCOUNT":     {"Account", "account"},
+		"SF_TEST_USER":        {"User", "user"},
+		"SF_TEST_PASSWORD":    {"Password", "password"},
+		"SF_TEST_ROLE":        {"Role", "role"},
+		"SF_TEST_HOST":        {"Host", "host"},
+		"SF_TEST_PORT":        {"Port", "8080"},
+		"SF_TEST_PROTOCOL":    {"Protocol", "http"},
+		"SF_TEST_WAREHOUSE":   {"Warehouse", "warehouse"},
+		"SF_TEST_DATABASE":    {"Database", "database"},
+		"SF_TEST_REGION":      {"Region", "region"},
+		"SF_TEST_PASSCODE":    {"Passcode", "passcode"},
+		"SF_TEST_SCHEMA":      {"Schema", "schema"},
+		"SF_TEST_APPLICATION": {"Application", "application"},
+	}
+	var properties = make([]*ConfigParam, len(envMap))
+	//TODO: Do not use append - directly add the param to the map
+	i := 0
+	for key, ctv := range envMap {
+		os.Setenv(key, ctv.value)
+		cfgParam := ConfigParam{ctv.configParam, key, true}
+		properties[i] = &cfgParam
+		i++
+	}
+	defer func() {
+		for key := range envMap {
+			os.Unsetenv(key)
+		}
+	}()
+
+	cfg, err := GetConfigFromEnv(properties)
+	if err != nil {
+		t.Errorf("unable to parse env variables to Config, err: %v", err)
+	}
+
+	err = checkConfig(*cfg, envMap)
+	if err != nil {
+		t.Error(err)
+	}
+}
+
+func checkConfig(cfg Config, envMap map[string]configParamToValue) error {
+	v := reflect.ValueOf(cfg)
+	typeOfCfg := v.Type()
+	cfgValues := make(map[string]interface{}, v.NumField())
+	for i := 0; i < v.NumField(); i++ {
+		cfgValues[typeOfCfg.Field(i).Name] = v.Field(i).Interface()
+	}
+
+	var errorsStr string
```

Comment:
Let me suggest another solution - maybe create a slice of errors here, append all errors, and in invoking function use this array in `t.Error(errors)`/`t.Error(strings.Join(errors, '\n')`? What do you think?

Reply:
That ok for me :)